libraries/rush-lib/src/logic/operations/ShellOperationRunner.ts

```diff
@@ -35,11 +35,14 @@ import { CollatedTerminalProvider } from '../../utilities/CollatedTerminalProvid
 import { RushConstants } from '../RushConstants';
 import { EnvironmentConfiguration } from '../../api/EnvironmentConfiguration';
 import { OperationMetadataManager } from './OperationMetadataManager';
+import { RunnerWatcher } from './RunnerWatcher';
+import { CobuildLock, ICobuildCompletedState } from '../cobuild/CobuildLock';
```

Comment:
The responsibilities of `ShellOperationRunner` are intended to be "perform this operation by invoking the specified shell command"; no more, no less. Interaction with the build cache already doesn't belong in this class (at minimum it should be a wrapper that takes an underlying `IOperationRunner` that does the build in the event that it can't replay from cache).
Cobuilds should likewise be a wrapper, so that they can be used with runners that aren't `ShellOperationRunner`.

Reply:
Thanks David for your valuable suggestions! Let me see how to refactor the part of logic 👀

Reply:
Hi @dmichon-msft , I've refactored `ShellOperationRunner`, and extracted all build-cache, cobuild related logic into a individual `CacheableOperationRunnerPlugin`. Please kindly take a look and i would like to hear your thoughts.